cmd/util/exit.go

```diff
@@ -0,0 +1,35 @@
+/*
+Copyright 2021 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"context"
+	"errors"
+	"runtime"
+)
+
+// GracefullyExit sets the exit code to 1 if the error is not a context.Canceled
+// error. It then calls runtime.Goexit() which means that all defer calls
+// will finish before exiting the command. This allows us to cleanup any
+// residual resources when exiting (even on error).
+func GracefullyExit(err error) {
+	if (err != nil) && !errors.Is(err, context.Canceled) {
+		errorExitCodeChannel <- 1 // Indicate that there was an error
+	}
+
+	runtime.Goexit() // Do soft exit (handle all defers)
```

Comment:
question: given the below:

```console
$ go doc runtime Goexit
package runtime // import "runtime"

func Goexit()
    Goexit terminates the goroutine that calls it. No other goroutine is
    affected. Goexit runs all deferred calls before terminating the goroutine.
    Because Goexit is not a panic, any recover calls in those deferred functions
    will return nil.

    Calling Goexit from the main goroutine terminates that goroutine without
    func main returning. Since func main has not returned, the program continues
    execution of other goroutines. If all other goroutines exit, the program
    crashes.
```

Specifically, it seems we always call `GracefullyExit` in the main goroutine, which would mean that if there's another goroutine running we could end up causing a crash, because of this line:

> Since func main has not returned, the program continues execution of other goroutines. If all other goroutines exit, the program crashes.

This seems like an undesirable behaviour to me, and could actually lead to more crashes if there _are_ other goroutines running, or else could lead to the program hanging if one of those goroutines doesn't exit.

Is this an issue? I think at the very least we should be documenting this in the doc comment for GracefullyExit, but I'm concerned that Goexit might not be what we want here.

Reply:
Interesting, we probably want all goroutines to exit gracefully.
This however is probably not taken into account everywhere (each go routine should stop if the context is canceled). Additionally, like you describe, the main thread should then wait for all these goroutines to finish -> this is probably not an easy task, I'll have another look and will let you know if I can find some more details on how to fix this.
The current solution will call os.Exit(...) when the outer exit defer is called, but it will not exit in case there is no errorcode in the errorcode channel. To prevent crashes, we can call os.Exit(0) if there is no errorcode in the errorcode channel.

Reply:
Just had a look at how goroutines are being awaited. I have found some examples that are being awaited and some that are not.
- closing a channel: https://github.com/jetstack/cert-manager/blob/c92c8c8f1323172b9959b66820d99f41e4eec0af/pkg/webhook/server/server.go#L233
- WaitGroup: https://github.com/jetstack/cert-manager/blob/c92c8c8f1323172b9959b66820d99f41e4eec0af/cmd/controller/app/controller.go#L126

I also found some functions that I think are not awaited properly:
https://github.com/jetstack/cert-manager/blob/c92c8c8f1323172b9959b66820d99f41e4eec0af/pkg/webhook/authority/authority.go#L212-L222
https://github.com/jetstack/cert-manager/blob/c92c8c8f1323172b9959b66820d99f41e4eec0af/cmd/acmesolver/app/app.go#L40-L45
https://github.com/jetstack/cert-manager/blob/c92c8c8f1323172b9959b66820d99f41e4eec0af/pkg/webhook/server/tls/dynamic_source.go#L63-L94

I think we should change these non-awaited goroutines to be awaited, as this behavior is not correct!
The go main function, on exit, will cause all these goroutines to exit and the cleanup functions to fail.

Reply:
see https://github.com/jetstack/cert-manager/pull/4243